### PR TITLE
F32 p9 p1

### PR DESCRIPTION
**Pull request by munroesj52**
Update vec_f32_ppc.h to use POWER9 test date class instructions.

### CODE REVIEW

File: src/vec_f32_ppc.h [hunk 1]
```diff
@@ -73,6 +97,101 @@
...
+ * The following code example uses functions from this header to
+ * address the POSIX requirements for special values input to
+ * for a vectorized sinf():
+ * \code
+vf32_t
+test_vec_sinf32 (vf32_t value)
+{
+  const vf32_t vec_f0 = { 0.0, 0.0, 0.0, 0.0 };
+  const vui32_t vec_f32_qnan =
+    { 0x7f800001, 0x7fc00000, 0x7fc00000, 0x7fc00000 };
```
Comment (ThinkOpenly) [inline src/vec_f32_ppc.h]: (nit) curious if these constants have mnemonics

File: src/vec_f32_ppc.h [hunk 2]
```diff
@@ -73,6 +97,101 @@
...
+      // replace non-normal input values with safe values.
+      vf32_t safeval = vec_sel (vec_f0, value, normmask);
+      // body of vec_sin(safeval) computation elided for this example.
+    }
+  else
+    result = value;
+
+  // merge non-normal input values back into result
+  result = vec_sel (value, result, normmask);
+  // Inf input value elements return quite-nan
```
Comment (ThinkOpenly) [inline src/vec_f32_ppc.h]: s/quite/quiet/ ?

File: src/vec_f32_ppc.h [hunk 3]
```diff
@@ -73,6 +97,101 @@
...
+    }
+  else
+    result = value;
+
+  // merge non-finite input values back into result
+  result = vec_sel (value, result, finitemask);
+  // Set +-0.0 input elements to exactly 1.0 in result
+  zeromask = vec_iszerof32 (value);
+  result = vec_sel (result, vec_f1, zeromask);
+  // Set Inf input elements to quite-nan in result
```
Comment (ThinkOpenly) [inline src/vec_f32_ppc.h]: s/quite/quiet/

File: src/vec_f32_ppc.h [hunk 4]
```diff
@@ -282,21 +477,73 @@ vec_all_issubnormalf32 (vf32_t vf32)
...
+ *  \note This function will not raise VXSNAN or VXVC (FE_INVALID)
+ *  exceptions. A normal float compare can.
+ *
+ *  |processor|Latency|Throughput|
+ *  |--------:|:-----:|:---------|
+ *  |power8   | 4-20  | 2/cycle  |
+ *  |power9   |   6   | 1/cycle  |
+ *
+ *  \note This function will not raise VXSNAN or VXVC (FE_INVALID)
+ *  exceptions. A normal __float128 compare can.
```
Comment (ThinkOpenly) [inline src/vec_f32_ppc.h]: This note is a duplicate of the note just 8 lines up.

**Comment by munroesj52** [comment]
On Thu, 2018-10-18 at 06:42 -0700, Paul Clarke wrote:
> @ThinkOpenly approved this pull request.
> global nit: many instances where missing space between called
> function name and opening parenthesis.
> In src/vec_f32_ppc.h:
> > + * - If the input <I>value</I> is NaN then return a NaN.
> + * - If the input <I>value</I> is +-0.0 then return <I>value</I>.
> + * - If the input <I>value</I> is subnormal then return
> <I>value</I>.
> + * - If the input <I>value</I> is +-Inf then return a NaN.
> + * - Otherwise compute and return sin(value).
> + *
> + * The following code example uses functions from this header to
> + * address the POSIX requirements for special values input to
> + * for a vectorized sinf():
> + * \code
> +vf32_t
> +test_vec_sinf32 (vf32_t value)
> +{
> +  const vf32_t vec_f0 = { 0.0, 0.0, 0.0, 0.0 };
> +  const vui32_t vec_f32_qnan =
> +    { 0x7f800001, 0x7fc00000, 0x7fc00000, 0x7fc00000 };
> (nit) curious if these constants have mnemonics
That would be nice but no.

This should be a builtin or at least a static const in altivec.h. But
the last time I suggested that, it was rejected.

Your windmill, Don Quixote ...

> In src/vec_f32_ppc.h:
> > +  vf32_t result;
> +  vb32_t normmask, infmask;
> +
> +  normmask = vec_isnormalf32 (value);
> +  if (vec_any_isnormalf32 (value))
> +    {
> +      // replace non-normal input values with safe values.
> +      vf32_t safeval = vec_sel (vec_f0, value, normmask);
> +      // body of vec_sin(safeval) computation elided for this
> example.
> +    }
> +  else
> +    result = value;
> +
> +  // merge non-normal input values back into result
> +  result = vec_sel (value, result, normmask);
> +  // Inf input value elements return quite-nan
> s/quite/quiet/ ?
Yes

> In src/vec_f32_ppc.h:
> > +  finitemask = vec_isfinitef32 (value);
> +  if (vec_any_isfinitef32 (value))
> +    {
> +      // replace non-finite input values with safe values
> +      vf32_t safeval = vec_sel (vec_f0, value, finitemask);
> +      // body of vec_sin(safeval) computation elided for this
> example
> +    }
> +  else
> +    result = value;
> +
> +  // merge non-finite input values back into result
> +  result = vec_sel (value, result, finitemask);
> +  // Set +-0.0 input elements to exactly 1.0 in result
> +  zeromask = vec_iszerof32 (value);
> +  result = vec_sel (result, vec_f1, zeromask);
> +  // Set Inf input elements to quite-nan in result
> s/quite/quiet/
Yes 

> In src/vec_f32_ppc.h:
> > + *
> + *  A IEEE Binary32 finite value has an exponent between 0x000 and
> + *  0x7f0 (a 0x7f8 indicates NaN or Inf).  The significand can be
> + *  any value.
> + *  The sign bit is ignored.
> + *
> + *  \note This function will not raise VXSNAN or VXVC (FE_INVALID)
> + *  exceptions. A normal float compare can.
> + *
> + *  |processor|Latency|Throughput|
> + *  |--------:|:-----:|:---------|
> + *  |power8   | 4-20  | 2/cycle  |
> + *  |power9   |   6   | 1/cycle  |
> + *
> + *  \note This function will not raise VXSNAN or VXVC (FE_INVALID)
> + *  exceptions. A normal __float128 compare can.
> This note is a duplicate of the note just 8 lines up.
Yes 

All will fixed in the next push

**Pull request merged**